src/threadpool.c

```diff
@@ -310,3 +332,44 @@ int uv_cancel(uv_req_t* req) {
 
   return uv__work_cancel(loop, req, wreq);
 }
+
+
+static void report(enum estage stage, int lock) {
+  QUEUE* q;
+  uv_queue_stats_t* s;
+  int length_;
+  int threads_;
+
+  if (lock) uv_mutex_lock(&mutex);
+
+  if (!QUEUE_EMPTY(&stats)) {
+    length_ = wq_length;
+    threads_ = idle_threads;
+
+    QUEUE_FOREACH(q, &stats) {
+      s = QUEUE_DATA(q, struct uv_queue_stats_s, q);
+      switch(stage) {
+        case SUBMIT: s->submit_cb(length_, threads_, s->data); break;
+        case START: s->start_cb(length_, threads_, s->data); break;
+        case DONE: s->done_cb(length_, threads_, s->data); break;
```

Comment:
The callbacks shouldn't be made with the lock held, should they?

Stylistic: libuv doesn't use everything-on-one-line style.  (Same goes for the `if` statements.)

Reply:
The only reason to release would be to allow the user to make calls that mutate the state of the locked queue, but if if the queue being iterated is mutated, it will break iteration.

Isn't this an issue? Is there a better way to deal with this? I wanted to allow libuv to be reentrant, and it was in the first version that only allowed a single callback but with the more complex data structure that doesn't seem to be possible any more.